parser/scanner.go

```diff
@@ -162,12 +165,21 @@ func (s *Scanner) EatRegexp(re *regexp.Regexp, match *Scanner, captures []Scanne
 
 // - stringSource
 
+func (s stringSource) stripSource(offset, size int) source {
+	s.origin = s.slice(offset, size)
+	return s
+}
+
 func (s stringSource) length() int {
-	return len(*s.origin)
+	return len(s.origin)
 }
 
 func (s stringSource) slice(i, length int) string {
-	return (*s.origin)[i : i+length]
+	// Since offset and length based on the original origin string, so they might be out of range
+	if i < 0 || i+length < 0 || i > len(s.origin) || i+length > len(s.origin) {
+		return s.origin
```

Comment:
This is also probably necessary, but again based on incorrect usage. And again, it can wait till later.